geode-core/src/integrationTest/java/org/apache/geode/internal/cache/event/DistributedEventTrackerIntegrationTest.java

```diff
@@ -0,0 +1,146 @@
+
+import org.apache.geode.cache.Operation;
+import org.apache.geode.cache.partition.PartitionRegionHelper;
+import org.apache.geode.distributed.internal.InternalDistributedSystem;
+import org.apache.geode.internal.cache.BucketRegion;
+import org.apache.geode.internal.cache.DistributedRegion;
+import org.apache.geode.internal.cache.EntryEventImpl;
+import org.apache.geode.internal.cache.EventID;
+import org.apache.geode.internal.cache.LocalRegion;
+import org.apache.geode.internal.cache.PartitionedRegion;
+import org.apache.geode.internal.cache.tier.sockets.ClientProxyMembershipID;
+import org.apache.geode.test.junit.categories.RegionsTest;
+import org.apache.geode.test.junit.rules.ServerStarterRule;
+import org.apache.geode.test.junit.rules.serializable.SerializableTestName;
+import org.apache.geode.test.junit.runners.CategoryWithParameterizedRunnerFactory;
+
+@RunWith(Parameterized.class)
+@Parameterized.UseParametersRunnerFactory(CategoryWithParameterizedRunnerFactory.class)
+@Category({RegionsTest.class})
+public class DistributedEventTrackerIntegrationTest {
+
+  @Parameterized.Parameters(name = "possibleDuplicate={0}")
+  public static Collection booleans() {
+    return Arrays.asList(true, false);
+  }
+
+  @Parameterized.Parameter
+  public boolean possibleDuplicate;
+
+  @Rule
+  public SerializableTestName testName = new SerializableTestName();
+
+  @Rule
+  public ServerStarterRule server = new ServerStarterRule().withNoCacheServer().withLogFile();
+
+  @Before
+  public void setUp() throws Exception {
+    server.startServer();
+  }
+
+  @Test
+  public void testHasSeenEventReplicatedRegion() throws IOException {
+    // Create the replicated region
+    DistributedRegion region = (DistributedRegion) server.createRegion(REPLICATE,
+        testName.getMethodName().substring(0, testName.getMethodName().indexOf('[')));
+
+    // Invoke hasSeenEvent and verify results
+    invokeHasSeenEventAndVerifyResults(region);
+  }
+
+  @Test
+  public void testHasSeenEventPartitionedRegion() throws IOException {
+    // Create the partitioned region
+    PartitionedRegion region = (PartitionedRegion) server.createRegion(PARTITION,
+        testName.getMethodName().substring(0, testName.getMethodName().indexOf('[')));
+
+    // Assign buckets so that the BucketRegions and their EventTrackers are created
+    PartitionRegionHelper.assignBucketsToPartitions(region);
+
+    // Get a BucketRegion
+    BucketRegion br = region.getBucketRegion(0);
+
+    // Invoke hasSeenEvent and verify results
+    invokeHasSeenEventAndVerifyResults(br);
+  }
+
+  private void invokeHasSeenEventAndVerifyResults(DistributedRegion region)
+      throws IOException {
+    // Record an event with a high sequence number
+    recordHighSequenceNumberEvent(region);
+
+    // Create an event with a lower sequence number and possibleDuplicate set appropriately
+    EntryEventImpl event =
+        createEvent(region, new EventID(new byte[0], 1l, 0l));
```

Comment:
Since the `threadId` used in the constructor for the `EventID` here needs to be the same as that used in the constructor for the `EventID` in `recordHighSequenceNumberEvent()`, it might be best to extract the value to a constant. Also, very minor nitpick, but IntelliJ creates a warning for me that it's better to use upper-case L to denote long values, so it seems like a simple thing to fix here too.

Reply:
Done. Thanks.